### PR TITLE
fix: add support for node-like module resolution

### LINKED ISSUE #4726: 🐛 BUG: onResolve error from bundling module resultion when importing wasm modules from npm packages
### Which Cloudflare product(s) does this pertain to?

Wrangler core

### What version(s) of the tool(s) are you using?

3.22.3

### What version of Node are you using?

20.10.0

### What operating system are you using?

Mac

### Describe the Bug

If I try to import a wasm module from an npm package in my worker using the node module resolution, for example:

```
import JPEG_DEC_WASM from '@jsquash/jpeg/codec/dec/mozjpeg_dec.wasm';
```

I get the following error when running `wrangler dev`:
```
✘ [ERROR] ENOENT: no such file or directory, open '/Users/foo/bar/src/@jsquash/jpeg/codec/dec/mozjpeg_dec.wasm' [plugin wrangler-module-collector]

    src/jsquash-transform.ts:3:26:
      3 │ import JPEG_DEC_WASM from '@jsquash/jpeg/codec/dec/mozjpeg_dec.wasm';
        ╵                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  This error came from the "onResolve" callback registered here:

    node_modules/.pnpm/wrangler@3.22.3_patch_hash=whuv2ih5kundyyu7rtaqweiv2a/node_modules/wrangler/wrangler-dist/cli.js:123256:19:
      123256 │             build5.onResolve(
             ╵                    ~~~~~~~~~

    at /Users/foo/bar/node_modules/.pnpm/wrangler@3.22.3_patch_hash=whuv2ih5kundyyu7rtaqweiv2a/node_modules/wrangler/wrangler-dist/cli.js:123256:20
    at Array.forEach (<anonymous>)
    at /Users/foo/bar/node_modules/.pnpm/wrangler@3.22.3_patch_hash=whuv2ih5kundyyu7rtaqweiv2a/node_modules/wrangler/wrangler-dist/cli.js:123255:22
    at Array.forEach (<anonymous>)
```

The only way I was able to get this to work was by using a relative path to the module rather than node module resolution:
```
import JPEG_DEC_WASM from '../node_modules/@jsquash/jpeg/codec/dec/mozjpeg_dec.wasm';
```

The module resolution should work the same for both js/ts and wasm files.

### Please provide a link to a minimal reproduction

_No response_

### Please provide any relevant error logs

_No response_

### PR DESCRIPTION
**Pull request by Cherry**
Fixes #4726

**What this PR solves / how to test:**

Similar to https://github.com/cloudflare/workers-sdk/pull/4135, this extends the module collection plugin to support standard node-like resolution of file paths.

This allows you to do things like:
```js
import JPEG_DEC_WASM from '@jsquash/jpeg/codec/dec/mozjpeg_dec.wasm';
```
even when the package doesn't explicitly export that file in its `exports` map.

This uses [`resolve`](https://www.npmjs.com/package/resolve), a package that was already in the dependency tree, and used throughout the JS ecosystem everywhere such as in `vite`, `jest`, `eslint-plugin-import`, etc.

**Author has addressed the following:**

- Tests
  - [x] Included
  - [ ] Not necessary because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [x] Included
  - [ ] Not necessary because:
- Associated docs
  - [ ] Issue(s)/PR(s):
  - [ ] Not necessary because:

**Note for PR author:**

We want to celebrate and highlight awesome PR review! If you think this PR received a particularly high-caliber review, please assign it the label `highlight pr review` so future reviewers can take inspiration and learn from it.

### CODE REVIEW

File: .changeset/sour-plums-pull.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+---
+"wrangler": patch
+---
+
+feat: resolve imports in a more node-like fashion, greatly increasing developer experience.
```
Comment (petebacondarwin) [inline .changeset/sour-plums-pull.md:5]:
```suggestion
feat: resolve imports in a more node-like fashion for package that do not declare exports
```

File: .changeset/sour-plums-pull.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+Previously, trying to import a file that wasn't explicitly exported from a module would result in an error, but now, better attempts are made to resolve the import using node's module resolution algorithm. It's now possible to do things like this:
+
+```js
+import JPEG_DEC_WASM from "@jsquash/jpeg/codec/dec/mozjpeg_dec.wasm";
+```
+
+This works even if the `mozjpeg_dec.wasm` file isn't explicitly exported from the `@jsquash/jpeg` module.
```
Comment (petebacondarwin) [inline .changeset/sour-plums-pull.md:13]:
```suggestion
This works even if the `mozjpeg_dec.wasm` file isn't explicitly exported from the `@jsquash/jpeg` module.

Fixes #4726 
```

**Review by petebacondarwin — APPROVED** [review]
LGTM - thanks for the fix.
Would you mind adjusting the changeset?

**Comment by Cherry** [comment]
Thanks @petebacondarwin, updated as suggested.

**Review by petebacondarwin — APPROVED** [review]

File: .changeset/sour-plums-pull.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
+---
+"wrangler": patch
+---
+
+feat: resolve imports in a more node-like fashion for packages that do not declare exports
```
Comment (petebacondarwin) [inline .changeset/sour-plums-pull.md:5]:
```suggestion
fix: resolve imports in a more node-like fashion for packages that do not declare exports
```

This is really a bug fix; and in any case if it was a feature we'd need a minor version bump here.

**Pull request merged**